src/io/flutter/inspector/HeapDisplay.java

```diff
@@ -0,0 +1,234 @@
+    public void update(List<IsolateObject> isolates) {
+      heapState.update(isolates);
+      for (HeapDisplay graph : graphs) {
+        // TODO(pq): stop storing points in the graph and just use the samples directly.
+        heapState.getSamples().forEach(sample -> graph.addMeasure(scale(heapState, sample)));
+      }
+      panels.forEach(panel -> SwingUtilities.invokeLater(panel::repaint));
+    }
+  }
+
+  /**
+   * A fixed-length list of captured samples.
+   */
+  public static class HeapSamples {
+    private final LinkedList<HeapSample> samples = new LinkedList<>();
+    private final int capacity;
+
+    HeapSamples(int size) {
+      this.capacity = size;
+    }
+
+    void add(HeapSample sample) {
+      if (samples.size() == capacity) {
+        samples.removeFirst();
+      }
+      samples.add(sample);
+    }
+
+    int size() {
+      return samples.size();
+    }
+
+    HeapSample get(int index) {
+      return samples.get(index);
+    }
+  }
+
+  public static class HeapState implements HeapListener {
+    // Running count of the max help (in bytes).
+    private int heapMaxInBytes;
+
+    private final HeapSamples samples;
+
+    HeapState(int sampleCapacity) {
+      samples = new HeapSamples(sampleCapacity);
+    }
+
+    public Iterable<HeapSample> getSamples() {
+      return samples.samples;
+    }
+
+    public int getMaxHeapInBytes() {
+      return heapMaxInBytes;
+    }
+
+    void addSample(HeapSample sample) {
+      samples.add(sample);
+    }
+
+    @Override
+    public void update(List<IsolateObject> isolates) {
+      int current = 0;
+      int total = 0;
+
+      for (IsolateObject isolate : isolates) {
+        for (HeapSpace heap : isolate.getHeaps()) {
+          current += heap.getUsed() + heap.getExternal();
+          total += heap.getCapacity() + heap.getExternal();
+        }
+      }
+
+      heapMaxInBytes = total;
+      addSample(new HeapSample(current, false));
+    }
+  }
+
+  private static final int TEN_MB = 1024 * 1024 * 10;
+  private static final double GRAPH_PIXEL_HIGHT = 16.0;
+
+  private static int scale(HeapState state, HeapSample sample) {
+    final double maxDataSize = (Math.round(state.getMaxHeapInBytes() / TEN_MB)) * TEN_MB + TEN_MB;
+    return (int)(GRAPH_PIXEL_HIGHT * sample.getBytes() / maxDataSize);
+  }
+
+  // TODO(pq): remove in favor of just using the HeapSamples directly.
+  static class IntCache {
+    final LinkedList<Integer> list = new LinkedList<>();
+    private final int capacity;
+
+    IntCache(int size) {
+      this.capacity = size;
+    }
+
+    void add(int v) {
+      if (list.size() == capacity) {
+        list.removeFirst();
+      }
+      list.add(v);
+    }
+
+    int size() {
+      return list.size();
+    }
+
+    int get(int index) {
+      return list.get(index);
+    }
+  }
+
+  private static final int PREFFERED_WIDTH = 60;
+  private static final int PREFFERED_HEIGHT = 16;
+
+  private static final Color GRAPH_COLOR = JBColor.LIGHT_GRAY;
```

Comment:
We'll want to try and take this from the current theme.

I see a `UIUtil.isUnderDarcula()` util method, and some general color manager classes; not sure which is the best option for getting a good foreground color.

Reply:
If I'm understanding JBColor right, this is done for us.  For example, light gray is declared like this:

     new JBColor(Gray._192, Gray._64);

where `Gray._192` is for regular presentation and `Gray._64` for dark.

Let me know if I'm missing something though!

Reply:
Ah, cool; I wasn't aware that `JBColor.LIGHT_GRAY` auto adjusted to the theme.

Reply:
(from digging into the source, we may want to prefer something like `UIUtil.getLabelForeground()`)

Reply:
Hmmmm.  I like the idea but that feels a little dark:

![image](https://user-images.githubusercontent.com/67586/37493504-5b855ca8-2862-11e8-8685-c09c76704d9f.png)

I'll poke around and see if there's something else a little more semantic than "gray".